cpp/src/arrow/io/caching.cc

```diff
@@ -139,8 +147,28 @@ struct ReadRangeCache::Impl {
   // Ordered by offset (so as to find a matching region by binary search)
   std::vector<RangeCacheEntry> entries;
 
-  // Add new entries, themselves ordered by offset
-  void AddEntries(std::vector<RangeCacheEntry> new_entries) {
+  virtual ~Impl() = default;
+
+  // Get the future corresponding to a range
+  virtual Future<std::shared_ptr<Buffer>> MaybeRead(RangeCacheEntry* entry) {
+    return entry->future;
+  }
+
+  // Make a cache entry for a range
+  virtual RangeCacheEntry MakeCacheEntry(const ReadRange& range) {
```

Comment:
You may make this `std::vector<RangeCacheEntry> MakeCacheEntries(const std::vector<ReadRange>&)` instead and you will issue one virtual call instead of N.